logstash-core/src/main/java/org/logstash/plugins/internal/Common.java

```diff
@@ -0,0 +1,126 @@
+package org.logstash.plugins.internal;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class Common {
+    public static class AddressState {
+        private final String address;
+        private final Set<InternalOutput> outputs = ConcurrentHashMap.newKeySet();
+        private volatile InternalInput input = null;
+
+        private AddressState(String address) {
+            this.address = address;
+            this.input = input;
+        }
+
+        public Set<InternalOutput> getOutputs() {
+            return outputs;
+        }
+
+        public InternalInput getInput() {
+            return input;
+        }
+    }
+
+    public static ConcurrentHashMap<String, AddressState> ADDRESS_STATES = new ConcurrentHashMap<>();
+
+    static class AddressesByRunState {
+        private final List<String> running;
+        private final List<String> notRunning;
+
+        AddressesByRunState() {
+            this.running = new ArrayList<>();
+            this.notRunning = new ArrayList<>();
+        }
+
+        public List<String> getRunning() {
+            return running;
+        }
+
+        public List<String> getNotRunning() {
+            return notRunning;
+        }
+    }
+
+    static AddressesByRunState addressesByRunState() {
+        AddressesByRunState addressesByRunState = new AddressesByRunState();
+        ADDRESS_STATES.forEach( (address, state) -> {
+            if (state.input != null && state.input.isRunning()) {
+                addressesByRunState.running.add(address);
+            } else {
+                addressesByRunState.notRunning.add(address);
+            }
+        });
+        return addressesByRunState;
+    }
+
+    public static void registerSender(Collection<String> addresses, InternalOutput output) {
+        addresses.forEach( address -> ADDRESS_STATES.compute(address, (a, state) -> {
+            if (state == null) {
+                state = new AddressState(address);
+            }
+            state.outputs.add(output);
+            if (state.input != null) {
+                output.updateAddressReceiver(address, state.input::internalReceive);
+            }
+            return state;
+        }));
+    }
+
+    public static void special(Collection<String> addresses, InternalOutput output) {
```

Comment:
`special` doesn't really tell me that this method unbinds the given output from the given addresses, and it may read clearer if we reverse the argument order:

~~~ diff
-    public static void special(Collection<String> addresses, InternalOutput output) {
+    public static void unregisterSender(final InternalOutput output, final Collection<String> addresses) {
~~~

Also, if one or more of the given `addresses` is not registered in `ADDRESS_STATES` , this will NPE dereferencing `state`.

Reply:
Sorry, this is from when I was experimenting with that JRuby not finding the method issue. I renamed it at one point to see if the previous name was somehow breaking JRuby.

I should mention, this is a WIP, and the intent was to get feedback more on general direction rather than specific code.

This is obviously a garbage method of course.